README.md
# mmaflash
Macromedia/Adobe Flash Tool

Build with GCC 3.2 (mingw special 20020817-1) from DEV-CPP 4.9.8.0 (2003).
Pure C and Win32API application, no GMO required
(like msvcrt.dll, .NET Framework, Visual Studio Redistributable and so on).
Should work anywhere starting from Windows 95/98.

Softpedia Editor's review (you can download binary "MMAFlash.exe" here):

http://www.softpedia.com/get/Internet/WEB-Design/Flash/Macromedia-Adobe-Flash-Tool.shtml

Software homepage (you can download binary "MMAFlash.exe" here):

http://systools.losthost.org/?misc#mmaflash

Licensed under the Apache License, Version 2.0 (the "License") - see LICENSE file.


resource/history.txt
v1.9 [2022.05.28]
- new method for Standalone Flash Player download (provided links for official and archived versions)
- add Flash file type FWS / CWS / ZWS (last two compressed) after version in information block
- some dialog text rephrase

v1.8 [2018.10.01]
- Adobe removes all older Standalone Flash Player files, only latest player available now
- add missing const to declaration of miniz array, no .data section, less executable size
- minor code cleanup

v1.7 [2017.03.15]
- added support for Flash 25
- fixed invalid HTTP request for latest available Standalone Player version detection
- popup menu for the list of available Standalone Players replaced with the combo box

v1.6 [2016.12.21]
- added support for Flash 24
- fixed version information block window styles (borders was invisible under Windows 98)
- fixed version information block height (last line with the add/replace player version was invisible)
- add minimize button - window can be minimized now
- add version resources
- minor code cleanup

v1.5 [2016.03.13]
This version changes aimed to make it more flexible to any possible future Flash releases.
- added support for Flash 21 and any other possible future versions if Adobe didn't
  change version increment method
- latest available Standalone Player version will be detected on the first usage of
  "Get player..." button per program run through official Adobe version check URL,
  but in case of any network errors version 21 will be used as the latest
- "Compress" option renamed to "(Re)Compress" because that's what it actually do -
  compress uncompressed Flash file or recompress already compressed Flash file
  (i.e. uncompress and compress it back again)
- minor code cleanup

v1.4 [2015.12.10]
- added support for Flash 20
- minor code cleanup

v1.3 [2015.09.22]
- added support for Flash 19
- source codes now available through GitHub repository: https://github.com/systoolz/mmaflash

v1.2 [2015.06.10]
- added support for Flash 18
- added "Get player..." button which allows to download any Standalone Player version
- version information block moved to the bottom and now displays more detailed output
- code cleanup, removed PNG routines from miniz

v1.1 [2015.04.26]
- added support for LZMA-compressed Flash files (only "Keep unchanged" mode)
- improved Flash and Standalone Player version comparsion
- required Standalone Flash Player version can be downloaded from application
- replaced zlib with miniz: reduced program size and added progress bar in cost of
  a bit worse compression (but still better than default one from Adobe)
- fixed manifest resource id (fixes window styles)
- fixed commandline routine
- set active link to the site
- source codes released to public

v1.0 [2014.05.27]
- first publice release


FlashKit.h
#ifndef __FLASHKIT_H
#define __FLASHKIT_H

// "FWS" (none compression)
#define FK_SIGN_FWS 0x00535746
// "CWS" (ZLIB compression)
#define FK_SIGN_CWS 0x00535743
// v1.1
// (!) ZWS file compression is permitted in SWF 13 or later only.
// "ZWS" (LZMA compression)
#define FK_SIGN_ZWS 0x0053575A
// "MZ"
#define FK_SIGN_EXE 0x5A4D
// overlay
#define FK_SIGN_END 0xFA123456

// get version
#define FK_GET_FVER(x) HIBYTE(HIWORD((x)))
// get sign
#define FK_GET_SIGN(x) ((x) & 0x00FFFFFF)

#pragma pack(push, 1)
typedef struct tagFLASHINFO {
  DWORD HeadSign; // FWS = ShockWaveFile; CWS = ShockWaveCompressed
  DWORD HeadSize; // FWS == filesize; CWS == unpacked filesize
  DWORD FileOffs; // Flash file offset (0 == from start, no overlay; this is also player size)
  DWORD FileSize; // Flash file length (same as headsize for FWS)
} FLASHINFO, *LPFLASHINFO;
#pragma pack(pop)

DWORD FK_GetRequiredPlayerVersion(DWORD dwFlash);
void FK_GetFileInfo(TCHAR *FileName, FLASHINFO *fi);
void FK_ExtractPlayer(TCHAR *fsource, TCHAR *filename);
void FK_HandleFile(HWND wnd, TCHAR *fsource, TCHAR *filename, TCHAR *fplayer, DWORD flags);

#endif


SysToolX.c
#include <windows.h>
#include <wininet.h>
#include "SysToolX.h"

void FreeMem(void *block) {
  if (block) {
    LocalFree(block);
  }
}

void *GetMem(DWORD dwSize) {
  return((void *) LocalAlloc(LPTR, dwSize));
}

void *GrowMem(void *block, DWORD dwSize) {
  if (!dwSize) {
    if (block) {
      FreeMem(block);
    }
    block = NULL;
  } else {
    if (block) {
      block = (void *) LocalReAlloc(block, dwSize, LMEM_MOVEABLE | LMEM_ZEROINIT);
    } else {
      block = GetMem(dwSize);
    }
  }
  return(block);
}

TCHAR *LangLoadString(UINT sid) {
TCHAR *res;
WORD *p;
HRSRC hr;
int i;
  res = NULL;
  hr = FindResource(NULL, MAKEINTRESOURCE(sid / 16 + 1), RT_STRING);
  p = hr ? (WORD *) LockResource(LoadResource(NULL, hr)) : NULL;
  if (p != NULL) {
    for (i = 0; i < (sid & 15); i++) {
      p += 1 + *p;
    }
    res = STR_ALLOC(*p);
    LoadString(NULL, sid, res, *p + 1);
  }
  return(res);
}

/* == COMMAD LINE ROUTINES ================================================= */

/* http://msdn.microsoft.com/en-us/library/17w5ykft.aspx */
#define CMD_PARSE_QUOTE 1
#define CMD_PARSE_COPY  2
void ParseCmdLine(TCHAR *p, TCHAR **argv, TCHAR *lpstr, DWORD *numargs, DWORD *numbytes) {
DWORD flags;
DWORD slashes;
  *numbytes = 0;
  *numargs = 1;
  if (argv) {
    *argv++ = lpstr;
  }
  flags = (*p == TEXT('\"')) ? 1 : 0;
  p += flags;
  while (
    (*p != TEXT('\0')) && (
      (  flags  && (*p != TEXT('\"'))) ||
      ((!flags) && (*p != TEXT('\t')) && (*p != TEXT(' ')))
    )
  ) {
    *numbytes += sizeof(TCHAR);
    if (lpstr) {
      *lpstr++ = *p;
    }
    p++;
  }
  *numbytes += sizeof(TCHAR);
  if (lpstr) {
    *lpstr++ = TEXT('\0');
  }
  flags &= (*p == TEXT('\"')) ? 1 : 0;
  p += flags;
  flags = 0;
  for ( ; ; ) {
    while ((*p == TEXT(' ')) || (*p == TEXT('\t'))) {
      ++p;
    }
    if (*p == TEXT('\0')) {
      break;
    }
    if (argv) {
      *argv++ = lpstr;
    }
    ++*numargs;
    for ( ; ; ) {
      flags |= CMD_PARSE_COPY;
      slashes = 0;
      while (*p == TEXT('\\')) {
        ++p;
        ++slashes;
      }
      if (*p == TEXT('\"')) {
        if ((slashes % 2) == 0) {
          if (flags & CMD_PARSE_QUOTE) {
            if (p[1] == TEXT('\"')) {
              p++;
            } else {
              flags ^= CMD_PARSE_COPY;
            }
          } else {
            flags ^= CMD_PARSE_COPY;
          }
          flags ^= CMD_PARSE_QUOTE;
        }
        slashes /= 2;
      }
      while (slashes--) {
        if (lpstr) {
          *lpstr++ = TEXT('\\');
        }
        *numbytes += sizeof(TCHAR);
      }
      if ((*p == TEXT('\0')) || ((!(flags & CMD_PARSE_QUOTE)) && ((*p == TEXT(' ')) || (*p == TEXT('\t'))))) {
        break;
      }
      if (flags & CMD_PARSE_COPY) {
        if (lpstr) {
          *lpstr++ = *p;
        }
        *numbytes += sizeof(TCHAR);
      }
      ++p;
    }
    if (lpstr) {
      *lpstr++ = TEXT('\0');
    }
    *numbytes += sizeof(TCHAR);
  }
}

TCHAR **GetCmdLineArgs(DWORD *argc) {
TCHAR **argv;
TCHAR *cmdline;
DWORD cmdsize;
TCHAR prgname[MAX_PATH];
  argv = NULL;
  if (argc) {
    cmdline = GetCommandLine();
    if ((!cmdline) || (!*cmdline)) {
      GetModuleFileName(NULL, prgname, MAX_PATH);
      cmdline = prgname;
    }
    ParseCmdLine(cmdline, NULL, NULL, argc, &cmdsize);
    argv = (TCHAR **) GetMem((*argc + 1) * sizeof(cmdline) + cmdsize);
    if (argv) {
      argv[*argc] = NULL;
      ParseCmdLine(
        cmdline, argv,
        (TCHAR *) (((BYTE *) argv) + (*argc + 1) * sizeof(cmdline)),
        argc, &cmdsize
      );
    }
  }
  return(argv);
}

/* == FILE AND DISK ROUTINES =============================================== */

TCHAR *GetFullFilePath(TCHAR *filename) {
TCHAR *result, *np;
int sz;
  sz = GetFullPathName(filename, 0, NULL, &np);
  result = STR_ALLOC(sz);
  if (result) {
    GetFullPathName(filename, sz + 1, result, &np);
    result[sz] = 0;
  }
  return(result);
}

TCHAR *GetWndText(HWND wnd) {
TCHAR *result;
int sz;
  sz = GetWindowTextLength(wnd);
  result = STR_ALLOC(sz);
  if (result) {
    GetWindowText(wnd, result, sz + 1);
    result[sz] = 0;
  }
  return(result);
}

TCHAR *OpenSaveDialog(HWND wnd, TCHAR *filemask, TCHAR *defext, int savedlg) {
OPENFILENAME ofn;
TCHAR filename[MAX_PATH], *result;
  result = NULL;
  filename[0] = 0;
  ZeroMemory(&ofn, sizeof(ofn));
  ofn.lStructSize = sizeof(ofn);
  ofn.hwndOwner   = wnd;
  ofn.nMaxFile    = MAX_PATH;
  ofn.lpstrFile   = filename;
  ofn.Flags       = OFN_HIDEREADONLY | OFN_PATHMUSTEXIST | OFN_OVERWRITEPROMPT; // | OFN_EXPLORER | OFN_FILEMUSTEXIST | OFN_NOCHANGEDIR
  /// OFN_ALLOWMULTISELECT
  // http://stackoverflow.com/questions/656655/getopenfilename-with-ofn-allowmultiselect-flag-set
  // http://support.microsoft.com/kb/131462
  ofn.lpstrFilter = filemask;
  ofn.lpstrDefExt = defext;
  if ((savedlg ? GetSaveFileName : GetOpenFileName)(&ofn)) {
    result = GetFullFilePath(ofn.lpstrFile);
  }
  return(result);
}

// v1.1
void URLOpenLink(HWND wnd, TCHAR *s) {
  CoInitialize(NULL);
  ShellExecute(wnd, NULL, s, NULL, NULL, SW_SHOWNORMAL);
  CoUninitialize();
}

int MsgBox(HWND wnd, TCHAR *lpText, UINT uType) {
int result;
TCHAR *s, *t;
  s = NULL;
  if (!HIWORD(lpText)) {
    s = LangLoadString(LOWORD(lpText));
  }
  t = GetWndText(wnd);
  result = MessageBox(wnd, s ? s : lpText, t, uType);
  if (t) { FreeMem(t); }
  if (s) { FreeMem(s); }
  return(result);
}

// http://blogs.msdn.com/b/oldnewthing/archive/2004/08/04/208005.aspx
void DialogEnableWindow(HWND hdlg, int idControl, BOOL bEnable) {
HWND hctl;
  hctl = GetDlgItem(hdlg, idControl);
  if ((bEnable == FALSE) && (hctl == GetFocus())) {
    SendMessage(hdlg, WM_NEXTDLGCTL, 0, FALSE);
  }
  EnableWindow(hctl, bEnable);
}

// PATCH: Windows 98 combobox fix (common controls version < 6)
// http://blogs.msdn.com/b/oldnewthing/archive/2006/03/10/548537.aspx
// ! http://vbnet.mvps.org/index.html?code/comboapi/comboheight.htm
void AdjustComboBoxHeight(HWND hWndCmbBox, DWORD MaxVisItems) {
RECT rc;
  GetWindowRect(hWndCmbBox, &rc);
  rc.right -= rc.left;
  ScreenToClient(GetParent(hWndCmbBox), (POINT *) &rc);
  rc.bottom = (MaxVisItems + 2) * SendMessage(hWndCmbBox, CB_GETITEMHEIGHT, 0, 0);
  MoveWindow(hWndCmbBox, rc.left, rc.top, rc.right, rc.bottom, TRUE);
  // PATCH: enable integral height, ComboBox must be created with CBS_NOINTEGRALHEIGHT
  SetWindowLong(hWndCmbBox, GWL_STYLE, (GetWindowLong(hWndCmbBox, GWL_STYLE) | CBS_NOINTEGRALHEIGHT) ^ CBS_NOINTEGRALHEIGHT);
}

DWORD GetFileVersionMS(TCHAR *filename) {
DWORD result;
HINSTANCE hl;
HGLOBAL hg;
HRSRC hr;
BYTE *data;
  result = 0;
  hl = LoadLibraryEx(filename, 0, LOAD_LIBRARY_AS_DATAFILE);
  if (hl) {
    // http://stackoverflow.com/questions/13941837/how-to-get-version-info-from-resources
    hr = FindResource(hl, MAKEINTRESOURCE(1), RT_VERSION);
    if (hr) {
      hg = LoadResource(hl, hr);
      if (hg) {
        data = (BYTE *) LockResource(hg);
        // http://www.csn.ul.ie/~caolan/publink/winresdump/winresdump/doc/resfmt.txt
        if (data) {
          result = MEM_MOVE(data, WORD);
          // "Each block of information is dword aligned." (c) link above
          while ((result > sizeof(VS_FIXEDFILEINFO)) && (MEM_MOVE(data, DWORD) != VS_FFI_SIGNATURE)) {
            result -= sizeof(DWORD);
            data   += sizeof(DWORD);
          }
          result = ((VS_FIXEDFILEINFO *) data)->dwFileVersionMS;
        } // data
      } // hg
    } // hr
    FreeLibrary(hl);
  } // hl
  return(result);
}


SysToolX.h
#ifndef __SYSTOOLX_H
#define __SYSTOOLX_H

#define MEM_MOVE(x, y) (*((y *)x))
#define STR_SIZE(x) ((lstrlen(x) + 1) * sizeof(TCHAR))
#define STR_ALLOC(x) ((TCHAR *) GetMem((x + 1) * sizeof(TCHAR)))

void FreeMem(void *block);
void *GetMem(DWORD dwSize);
void *GrowMem(void *block, DWORD dwSize);

TCHAR *LangLoadString(UINT sid);

TCHAR **GetCmdLineArgs(DWORD *argc);

TCHAR *GetFullFilePath(TCHAR *filename);
TCHAR *GetWndText(HWND wnd);

TCHAR *OpenSaveDialog(HWND wnd, TCHAR *filemask, TCHAR *defext, int savedlg);

// v1.1
void URLOpenLink(HWND wnd, TCHAR *s);

int MsgBox(HWND wnd, TCHAR *lpText, UINT uType);
void DialogEnableWindow(HWND hdlg, int idControl, BOOL bEnable);
void AdjustComboBoxHeight(HWND hWndCmbBox, DWORD MaxVisItems); // v1.7
DWORD GetFileVersionMS(TCHAR *filename);

// v1.5
BYTE *HTTPGetContent(TCHAR *url, DWORD *len);

#endif


resource/MMAFlash.h
#ifndef __MMAFLASH_H
#define __MMAFLASH_H

// dialog
#define IDD_DLG        100

// icon
#define IDI_ICN          1

// control
#define IDC_FLOPEN     100
#define IDC_GETPLAY    101
#define IDC_FSOURCE    102
#define IDC_FRADIO1    103
#define IDC_FRADIO2    104
#define IDC_FRADIO3    105
#define IDC_PRADIO1    106
#define IDC_PRADIO2    107
#define IDC_PRADIO3    108
#define IDC_EXTRACT    109
#define IDC_FPLAYER    110
#define IDC_FLSAVE     111
#define IDC_VERSION    112
#define IDC_HOMEPAGE   113
#define IDC_PROGRESS   114

// string
#define IDS_MSK_SWF         1
#define IDS_MSK_EXE         2
#define IDS_MSK_ALL         3

#define IDS_MSG_BADVERSION  4
#define IDS_MSG_BADPLAYER   5
#define IDS_MSG_BADFLASH    6
#define IDS_MSG_DONE        7
#define IDS_MSG_LZMA        8
#define IDS_MSG_LOCATION    9

#define IDS_VER_FLASHMOVIE 10
#define IDS_VER_PLAYERMREQ 11
#define IDS_VER_PLAYERFILE 12
#define IDS_VER_PLAYERFROM 13

#define IDS_PLAYER_UPDATER 14
#define IDS_PLAYER_ARCHIVE 15

#endif


FlashKit.c
#include <windows.h>
#include <commctrl.h>
#include "SysToolX.h"
#include "FlashKit.h"

// http://code.google.com/p/miniz/
// miniz flags
#define MINIZ_NO_STDIO
#define MINIZ_NO_TIME
#define MINIZ_NO_ARCHIVE_APIS
#define MINIZ_NO_ARCHIVE_WRITING_APIS
// miniz routines
#define malloc(x) GetMem((x))
#define free(x) FreeMem((x))
#define realloc(x,y) GrowMem((x),(y))
#define MINIZ_NO_ZLIB_APIS
#define MINIZ_NO_ZLIB_COMPATIBLE_NAMES
// remove assert()
#define NDEBUG
#include "miniz/miniz.c"
#undef NDEBUG

// FIXME: miniz stubs
void *memset(void *s, int c, size_t n) {
char *p;
  p = s;
  while (n--) {
    *(p++) = (char) c;
  }
  return(s);
}

// --------------------------------------------------------------------------

/*
// older version
...
{ 9,  9, 0}
{10, 10, 0}
// first block
{11, 10, 2}
{12, 10, 3}
// second block
{13, 11, 0}
{14, 11, 1}
{15, 11, 2}
{16, 11, 3}
{17, 11, 4}
{18, 11, 5}
{19, 11, 6}
{20, 11, 7}
{21, 11, 8}
{22, 11, 9}
// third block
{23, 12, 0}
{24, 13, 0}
...
*/

#define MAKEVERS(x,y) MAKELONG((y),(x))

// http://www.adobe.com/devnet/articles/flashplayer-air-feature-list.html
// http://sleepydesign.blogspot.com/2012/04/flash-swf-version-meaning.html
DWORD FK_GetRequiredPlayerVersion(DWORD dwFlash) {
DWORD result;
  // according to Adobe specification:
  // "ZWS file compression is permitted in SWF 13 or later only."
  if (FK_GET_SIGN(dwFlash) == FK_SIGN_ZWS) {
    dwFlash = max(FK_GET_FVER(dwFlash), 13);
  } else {
    dwFlash = FK_GET_FVER(dwFlash);
  }
  // fix warnings
  result = 0;
  // older version
  if (dwFlash <= 10) {
    result = MAKEVERS(dwFlash, 0);
  }
  // newer version
  // first block
  if ((dwFlash >= 11) && (dwFlash <= 12)) {
    result = MAKEVERS(10, dwFlash - 9);
  }
  // second block
  if ((dwFlash >= 13) && (dwFlash <= 22)) {
    result = MAKEVERS(11, dwFlash - 13);
  }
  // third block
  // should work for any future versions
  // if Adobe didn't change version increment method
  if (dwFlash >= 23) {
    result = MAKEVERS(dwFlash - 11, 0);
  }
  return(result);
}

void FK_GetFileInfo(TCHAR *FileName, FLASHINFO *fi) {
DWORD flag, dw, data;
HANDLE fl;
  flag = 0;
  ZeroMemory(fi, sizeof(*fi));
  fl = CreateFile(FileName, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, 0);
  if (fl != INVALID_HANDLE_VALUE) {
    fi->FileSize = GetFileSize(fl, NULL);
    ReadFile(fl, &data, 4, &dw, NULL);
    if (LOWORD(data) == FK_SIGN_EXE) {
      // FIXME: check for correct player executable
      if (1) {
        SetFilePointer(fl, fi->FileSize - 8, NULL, FILE_BEGIN);
        fi->FileOffs = fi->FileSize;
        data = 0;
        ReadFile(fl, &data, 4, &dw, NULL);
        // check attach
        if (data == FK_SIGN_END) {
          // read attach size
          ReadFile(fl, &fi->FileSize, 4, &dw, NULL);
          fi->FileOffs -= (8 + fi->FileSize);
          SetFilePointer(fl, fi->FileOffs, NULL, FILE_BEGIN);
          flag |= 4;
        }
        flag |= 2;
      }
    }
    if (flag&4) {
      flag ^= 4;
    } else {
      SetFilePointer(fl, 0, NULL, FILE_BEGIN);
    }
    // read attach signature
    ReadFile(fl, fi, 8, &dw, NULL);
    data = FK_GET_SIGN(fi->HeadSign);
    // v1.1: add ZWS
    if ((data == FK_SIGN_FWS) || (data == FK_SIGN_CWS) || (data == FK_SIGN_ZWS)) {
      flag |= 1;
    }
    CloseHandle(fl);
  }
  data = (flag&2) ? fi->FileOffs : 0;
  if (!(flag&1)) {
    ZeroMemory(fi, sizeof(*fi));
  }
  fi->FileOffs = data;
}

// v1.1
#pragma pack(push, 1)
typedef struct {
  HWND wnd;
  DWORD BlockPos;
  DWORD FileSize;
  BYTE *outbuf;
} MZCBDATA;
#pragma pack(pop)

// v1.1
static int tinfl_put_buf_func(const void *pBuf, int len, void *pUser) {
MZCBDATA *cd;
  cd = (MZCBDATA *) pUser;
  // prevent buffer overflow
  len = min(len, cd->FileSize - cd->BlockPos);
  // update handled block size
  cd->BlockPos += len;
  // prevent divide by 0
  if (cd->FileSize >= 100) {
    // update progress bar %
    SendMessage(cd->wnd, PBM_SETPOS, cd->BlockPos/(cd->FileSize/100), 0);
    // update window
    UpdateWindow(GetParent(cd->wnd));
  }
  // move data block to output buffer
  CopyMemory(cd->outbuf, pBuf, len);
  // move output buffer pointer
  cd->outbuf += len;
  // prevent buffer overflow
  return(cd->BlockPos < cd->FileSize);
}

BYTE *FK_UnpackFile(HWND wnd, BYTE *pbuf, DWORD *psize) {
DWORD usize;
BYTE *ubuf;
MZCBDATA cd;
  ubuf = pbuf;
  if ((*psize > 8) && ((MEM_MOVE(pbuf, DWORD) & 0x00FFFFFF) == FK_SIGN_CWS)) {
    usize = MEM_MOVE(&pbuf[4], DWORD);
    ubuf = (BYTE *) GetMem(usize);
    // memory allocated
    if (ubuf) {
      usize -= 8;
//      uncompress(&ubuf[8], &usize, &pbuf[8], *psize - 8);
      // v1.1
      cd.wnd = wnd;
      cd.BlockPos = 0;
      cd.FileSize = usize;
      cd.outbuf = &ubuf[8];
      usize = *psize - 8;
      tinfl_decompress_mem_to_callback(&pbuf[8], (size_t *) &usize, tinfl_put_buf_func, &cd, TINFL_FLAG_PARSE_ZLIB_HEADER);
      // ===
      MEM_MOVE(ubuf, DWORD) = (MEM_MOVE(pbuf, DWORD) & 0xFF000000) | FK_SIGN_FWS;
      *psize = MEM_MOVE(&pbuf[4], DWORD);
      MEM_MOVE(&ubuf[4], DWORD) = *psize;
      FreeMem(pbuf);
    } else {
      ubuf = pbuf;
    }
  }
  return(ubuf);
}

BYTE *FK_PackFile(HWND wnd, BYTE *ubuf, DWORD *usize) {
DWORD psize, us;
BYTE *pbuf;
MZCBDATA cd;
  pbuf = ubuf;
  if ((*usize > 8) && ((MEM_MOVE(ubuf, DWORD) & 0x00FFFFFF) == FK_SIGN_FWS)) {
    us = MEM_MOVE(&ubuf[4], DWORD);
    if (us > 8) {
      us -= 8;
//      psize = compressBound(us);
      // FIXME: copy/paste from miniz.c
      // saves about 2.5 Kb of executable with MINIZ_NO_ZLIB_* flags
      psize = max(128 + (us * 110) / 100, 128 + us + ((us / (31 * 1024)) + 1) * 5);
      pbuf = (BYTE *) GetMem(psize + 8);
      // memory allocated
      if (pbuf) {
//        compress2(&pbuf[8], &psize, &ubuf[8], us, Z_BEST_COMPRESSION); // Z_OK
        // v1.1
        cd.wnd = wnd;
        cd.BlockPos = 0;
        cd.FileSize = psize;
        cd.outbuf = &pbuf[8];
        // 0x3300 - Z_BEST_COMPRESSION + zlib flags
        tdefl_compress_mem_to_output(&ubuf[8], us, tinfl_put_buf_func, &cd, 0x3300);
        // fix output size
        psize = cd.BlockPos;
        // ===
        MEM_MOVE(pbuf, DWORD) = (MEM_MOVE(ubuf, DWORD) & 0xFF000000) | FK_SIGN_CWS;
        MEM_MOVE(&pbuf[4], DWORD) = MEM_MOVE(&ubuf[4], DWORD);
        FreeMem(ubuf);
        *usize = psize + 8;
      } else {
        pbuf = ubuf;
      }
    }
  }
  return(pbuf);
}

BYTE *FK_ReadBlock(TCHAR *filename, DWORD Offs, DWORD Size) {
BYTE *result;
HANDLE fl;
DWORD dw;
  result = NULL;
  // sanity check
  if (filename && Size) {
    fl = CreateFile(filename, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, 0);
    if (fl != INVALID_HANDLE_VALUE) {
      SetFilePointer(fl, Offs, NULL, FILE_BEGIN);
      result = (BYTE *) GetMem(Size);
      ReadFile(fl, result, Size, &dw, NULL);
      CloseHandle(fl);
    }
  }
  return(result);
}

void FK_ExtractPlayer(TCHAR *fsource, TCHAR *filename) {
FLASHINFO fs;
HANDLE fl;
DWORD dw;
BYTE *p;
  // sanity check
  if (fsource && filename) {
    // get source file information
    FK_GetFileInfo(fsource, &fs);
    if (fs.FileOffs) {
      p = FK_ReadBlock(fsource, 0, fs.FileOffs);
      if (p) {
        // create output file
        fl = CreateFile(filename, GENERIC_WRITE | GENERIC_READ, 0, NULL, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, 0);
        if (fl != INVALID_HANDLE_VALUE) {
          WriteFile(fl, p, fs.FileOffs, &dw, NULL);
          CloseHandle(fl);
        } // CreateFile()
        FreeMem(p);
      } // p
    } // offs
  } // schk
}

void FK_HandleFile(HWND wnd, TCHAR *fsource, TCHAR *filename, TCHAR *fplayer, DWORD flags) {
FLASHINFO fs, fp;
DWORD dw, sz;
HANDLE fl;
BYTE *p;
  if (flags) {
    // get source file information
    FK_GetFileInfo(fsource, &fs);
    if (fs.FileSize) {
      // create output file
      fl = CreateFile(filename, GENERIC_WRITE | GENERIC_READ, 0, NULL, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, 0);
      if (fl != INVALID_HANDLE_VALUE) {
        p = NULL;
        sz = 0;
        fp.FileOffs = 0;
        // handle player
        switch (HIWORD(flags)) {
          // keep player
          case 0:
            // player exists in input file - move player to output file
            if (fs.FileOffs) {
              p = FK_ReadBlock(fsource, 0, fs.FileOffs);
              sz = fs.FileOffs;
            }
          // add/replace player file from fplayer
          case 2:
            // check that we don't go from case 0
            if (HIWORD(flags) == 2) {
              FK_GetFileInfo(fplayer, &fp);
              p = FK_ReadBlock(fplayer, 0, fp.FileOffs);
              sz = fp.FileOffs;
            }
            if (p) {
              WriteFile(fl, p, sz, &dw, NULL);
              FreeMem(p);
              // flag
              fp.FileOffs = sz;
            }
            break;
        }
        // handle flash
        sz = fs.FileSize;
        p = FK_ReadBlock(fsource, fs.FileOffs, sz);
        switch (LOWORD(flags)) {
          // decompress to FWS
          case 1:
          // compress to CWS
          case 2:
            // if compress selected for already compressed file
            // we should recompress it (uncompress and compress back)
            p = FK_UnpackFile(wnd, p, &sz);
            // check that we don't go from case 1
            if (LOWORD(flags) == 2) {
              p = FK_PackFile(wnd, p, &sz);
            }
            break;
        }
        if (p) {
          WriteFile(fl, p, sz, &dw, NULL);
          FreeMem(p);
        }
        // if there is a player - add footer
        if (fp.FileOffs) {
          fp.FileOffs = FK_SIGN_END;
          WriteFile(fl, &fp.FileOffs, 4, &dw, NULL);
          WriteFile(fl, &sz, 4, &dw, NULL);
        }
        CloseHandle(fl);
      } // CreateFile()
    } // fs.FileSize
  } // flags
}
